File: partial-capture-total-mismatch/node/sync-partial-capture.js
/**
 * Line up a WooCommerce order total with a partial Stripe capture.
 *
 * A store on manual capture can capture less than the full authorized amount, for a
 * split shipment, a stock shortfall, or a deliberate partial charge. Stripe's
 * PaymentIntent then shows the real amount taken in `amount_received`, but the
 * WooCommerce order was created with the original, larger total and nothing updates
 * it. The order overstates what the buyer actually paid. This walks recent paid
 * orders, reads the saved Stripe PaymentIntent id from order meta
 * `_stripe_intent_id` (falling back to `transaction_id`), and for any order whose
 * total is higher than what Stripe actually captured, corrects the order total to
 * match and adds a note explaining the change. Safe by default, dry run first.
 *
 * Guide: https://www.allanninal.dev/woocommerce/partial-capture-total-mismatch/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 14);
const MISMATCH_TOLERANCE_MINOR = Number(process.env.MISMATCH_TOLERANCE_MINOR || 1);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const PAID_STATUSES = new Set(["processing", "completed"]);

export function intentIdOf(order) {
  for (const meta of order.meta_data || []) {
    if (meta.key === "_stripe_intent_id" && meta.value) return meta.value;
  }
  const tid = order.transaction_id;
  return tid && tid.startsWith("pi_") ? tid : null;
}

export function orderTotalMinor(order) {
  // Works for two decimal currencies. Zero decimal currencies (JPY and friends)
  // have their own guide, since 50.00 is wrong for those.
  return Math.round(parseFloat(order.total) * 100);
}

export function capturedMinor(intent) {
  return intent.amount_received || 0;
}

export function toMajorStr(minor) {
  return (minor / 100).toFixed(2);
}

export function decide(order, intent, toleranceMinor = MISMATCH_TOLERANCE_MINOR) {
  if (!PAID_STATUSES.has(order.status)) return ["skip", "order not in a paid state"];
  if (!intent) return ["skip", "no Stripe PaymentIntent id on this order"];
  if (intent.status !== "succeeded" && intent.status !== "requires_capture") {
    return ["skip", "intent has no capture to compare yet"];
  }
  if ((intent.amount_capturable || 0) > 0) {
    return ["skip", "capture is still partial in progress, more may be captured"];
  }

  const orderMinor = orderTotalMinor(order);
  const chargedMinor = capturedMinor(intent);
  const drift = orderMinor - chargedMinor;

  if (Math.abs(drift) <= toleranceMinor) return ["ok", "order total matches what Stripe captured"];
  if (drift < 0) {
    return ["flag", `order total is lower than the Stripe charge (drift ${drift} minor units)`];
  }
  return [
    "fix",
    `only ${chargedMinor} of ${orderMinor} minor units was captured, order total should drop to match`,
  ];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function getIntent(intentId) {
  if (!intentId) return null;
  try {
    return await stripe.paymentIntents.retrieve(intentId);
  } catch {
    return null;
  }
}

async function* paidOrders() {
  const after = new Date(Date.now() - LOOKBACK_DAYS * 86400000).toISOString();
  let page = 1;
  while (true) {
    const batch = await woo(`/orders?status=processing,completed&after=${after}&per_page=50&page=${page}`);
    if (!batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function applyFix(order, intent, reason) {
  const newTotal = toMajorStr(capturedMinor(intent));
  await woo(`/orders/${order.id}`, { method: "PUT", body: JSON.stringify({ total: newTotal }) });
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Total corrected for a partial capture: ${reason}. Order total set to ` +
            `${newTotal} to match Stripe PaymentIntent ${intent.id}. Please review ` +
            `line items if this needs a refund too.`,
    }),
  });
}

async function flag(order, intent, reason) {
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Capture check failed: ${reason}. PaymentIntent ${intent.id}. ` +
            `Please review before shipping or refunding.`,
    }),
  });
}

export async function run() {
  let fixed = 0;
  let flagged = 0;
  for await (const order of paidOrders()) {
    const intent = await getIntent(intentIdOf(order));
    const [action, reason] = decide(order, intent);
    if (action === "skip") continue;
    if (action === "flag") {
      console.warn(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would flag" : "flagging"}`);
      if (!DRY_RUN) await flag(order, intent, reason);
      flagged++;
      continue;
    }
    console.log(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would fix" : "fixing"}`);
    if (!DRY_RUN) await applyFix(order, intent, reason);
    fixed++;
  }
  console.log(
    `Done. ${fixed} order(s) ${DRY_RUN ? "to fix" : "fixed"}, ` +
    `${flagged} order(s) ${DRY_RUN ? "to flag" : "flagged"}.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: partial-capture-total-mismatch/python/sync_partial_capture.py
"""Line up a WooCommerce order total with a partial Stripe capture.

A store on manual capture can capture less than the full authorized amount, for a
split shipment, a stock shortfall, or a deliberate partial charge. Stripe's
PaymentIntent then shows the real amount taken in `amount_received`, but the
WooCommerce order was created with the original, larger total and nothing updates
it. The order overstates what the buyer actually paid. This walks recent paid
orders, reads the saved Stripe PaymentIntent id from order meta
`_stripe_intent_id` (falling back to `transaction_id`), and for any order whose
total is higher than what Stripe actually captured, corrects the order total to
match and adds a note explaining the change. Safe by default, dry run first.
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("sync_partial_capture")

stripe.api_key = os.environ["STRIPE_SECRET_KEY"]
WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "14"))
MISMATCH_TOLERANCE_MINOR = int(os.environ.get("MISMATCH_TOLERANCE_MINOR", "1"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

PAID_STATUSES = {"processing", "completed"}


def intent_id_of(order):
    """The saved Stripe PaymentIntent id, from meta _stripe_intent_id or transaction_id."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == "_stripe_intent_id" and meta.get("value"):
            return meta["value"]
    tid = order.get("transaction_id")
    return tid if tid and tid.startswith("pi_") else None


def order_total_minor(order):
    """Order total in minor units (cents). Two decimal currencies only; zero decimal
    currencies such as JPY have their own guide, since round(x * 100) is wrong there."""
    return round(float(order["total"]) * 100)


def captured_minor(intent):
    """What Stripe actually captured for this intent, in minor units."""
    return intent.get("amount_received", 0)


def to_major_str(minor):
    """Minor units back to a two decimal string suitable for a WooCommerce total."""
    return f"{minor / 100:.2f}"


def decide(order, intent, tolerance_minor=MISMATCH_TOLERANCE_MINOR):
    """Pure decision: given an order and its Stripe PaymentIntent, decide whether the
    order total needs to be brought down to the amount Stripe actually captured.
    No I/O here, so this is fully unit testable.
    """
    if order["status"] not in PAID_STATUSES:
        return ("skip", "order not in a paid state")
    if intent is None:
        return ("skip", "no Stripe PaymentIntent id on this order")
    if intent.get("status") not in ("succeeded", "requires_capture"):
        return ("skip", "intent has no capture to compare yet")
    if intent.get("amount_capturable", 0) > 0:
        return ("skip", "capture is still partial in progress, more may be captured")

    order_minor = order_total_minor(order)
    charged_minor = captured_minor(intent)
    drift = order_minor - charged_minor

    if abs(drift) <= tolerance_minor:
        return ("ok", "order total matches what Stripe captured")
    if drift < 0:
        # Order total is lower than what Stripe took. That is an overcharge, not a
        # partial capture, and deserves its own careful look rather than an auto-fix.
        return ("flag", f"order total is lower than the Stripe charge (drift {drift} minor units)")

    return (
        "fix",
        f"only {charged_minor} of {order_minor} minor units was captured, "
        f"order total should drop to match",
    )


def get_intent(intent_id):
    if not intent_id:
        return None
    try:
        return stripe.PaymentIntent.retrieve(intent_id)
    except stripe.error.InvalidRequestError:
        return None


def paid_orders():
    page = 1
    after = f"{__import__('datetime').date.today() - __import__('datetime').timedelta(days=LOOKBACK_DAYS)}T00:00:00"
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"status": "processing,completed", "after": after, "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def apply_fix(order, intent, reason):
    new_total = to_major_str(captured_minor(intent))
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}",
        json={"total": new_total},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": f"Total corrected for a partial capture: {reason}. "
                      f"Order total set to {new_total} to match Stripe PaymentIntent "
                      f"{intent['id']}. Please review line items if this needs a refund too."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def flag(order, intent, reason):
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": f"Capture check failed: {reason}. PaymentIntent "
                      f"{intent['id']}. Please review before shipping or refunding."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    fixed = 0
    flagged = 0
    for order in paid_orders():
        intent = get_intent(intent_id_of(order))
        action, reason = decide(order, intent)
        if action == "skip":
            continue
        if action == "flag":
            log.warning("Order %s: %s. %s", order["id"], reason, "would flag" if DRY_RUN else "flagging")
            if not DRY_RUN:
                flag(order, intent, reason)
            flagged += 1
            continue
        log.info("Order %s: %s. %s", order["id"], reason, "would fix" if DRY_RUN else "fixing")
        if not DRY_RUN:
            apply_fix(order, intent, reason)
        fixed += 1
    log.info(
        "Done. %d order(s) %s, %d order(s) %s.",
        fixed, "to fix" if DRY_RUN else "fixed",
        flagged, "to flag" if DRY_RUN else "flagged",
    )


if __name__ == "__main__":
    run()
